<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Count normalization | RNA-seq-analysis</title>
  <meta name="description" content="RNA-seq analysis workshop" />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Count normalization | RNA-seq-analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bookdown.org/ggiaever/RNA-seq-analysis/img/football22.png" />
  <meta property="og:description" content="RNA-seq analysis workshop" />
  <meta name="github-repo" content="gurinina/RNA-seq-analysis" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Count normalization | RNA-seq-analysis" />
  
  <meta name="twitter:description" content="RNA-seq analysis workshop" />
  <meta name="twitter:image" content="https://bookdown.org/ggiaever/RNA-seq-analysis/img/football22.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differential-gene-expression-dge-analysis-overview.html"/>
<link rel="next" href="quality-control.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/vis-9.1.0/vis-network.min.css" rel="stylesheet" />
<script src="libs/vis-9.1.0/vis-network.min.js"></script>
<script src="libs/visNetwork-binding-2.1.2/visNetwork.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="css/git.css" type="text/css" />
<link rel="stylesheet" href="css/toc.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>RNA-seq analysis workshop</a></li>
<li class="chapter" data-level="1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html"><i class="fa fa-check"></i><b>1</b> Differential gene expression (DGE) analysis overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#review-of-the-dataset"><i class="fa fa-check"></i><b>1.1</b> Review of the dataset</a></li>
<li class="chapter" data-level="1.2" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#setting-up"><i class="fa fa-check"></i><b>1.2</b> Setting up</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#loading-libraries"><i class="fa fa-check"></i><b>1.2.1</b> Loading libraries</a></li>
<li class="chapter" data-level="1.2.2" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#loading-data"><i class="fa fa-check"></i><b>1.2.2</b> Loading data</a></li>
<li class="chapter" data-level="1.2.3" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#viewing-data"><i class="fa fa-check"></i><b>1.2.3</b> Viewing data</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#differential-gene-expression-analysis-overview"><i class="fa fa-check"></i><b>1.3</b> Differential gene expression analysis overview</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#rna-seq-count-distribution"><i class="fa fa-check"></i><b>1.3.1</b> RNA-seq count distribution</a></li>
<li class="chapter" data-level="1.3.2" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#modeling-count-data"><i class="fa fa-check"></i><b>1.3.2</b> Modeling count data</a></li>
<li class="chapter" data-level="1.3.3" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#improving-mean-estimates-i.e.-reducing-variance-with-biological-replicates"><i class="fa fa-check"></i><b>1.3.3</b> Improving mean estimates (i.e. reducing variance) with biological replicates</a></li>
<li class="chapter" data-level="1.3.4" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#differential-expression-analysis-workflow"><i class="fa fa-check"></i><b>1.3.4</b> Differential expression analysis workflow</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="count-normalization.html"><a href="count-normalization.html"><i class="fa fa-check"></i><b>2</b> Count normalization</a>
<ul>
<li class="chapter" data-level="2.1" data-path="count-normalization.html"><a href="count-normalization.html#normalization"><i class="fa fa-check"></i><b>2.1</b> Normalization</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="count-normalization.html"><a href="count-normalization.html#common-normalization-methods"><i class="fa fa-check"></i><b>2.1.1</b> Common normalization methods</a></li>
<li class="chapter" data-level="2.1.2" data-path="count-normalization.html"><a href="count-normalization.html#rpkmfpkm-not-recommended"><i class="fa fa-check"></i><b>2.1.2</b> RPKM/FPKM (not recommended)</a></li>
<li class="chapter" data-level="2.1.3" data-path="count-normalization.html"><a href="count-normalization.html#deseq2-normalized-counts-median-of-ratios-method"><i class="fa fa-check"></i><b>2.1.3</b> DESeq2-normalized counts: Median of ratios method</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="count-normalization.html"><a href="count-normalization.html#geometric-mean"><i class="fa fa-check"></i><b>2.2</b> Geometric Mean</a></li>
<li class="chapter" data-level="2.3" data-path="count-normalization.html"><a href="count-normalization.html#count-normalization-of-mov10-dataset"><i class="fa fa-check"></i><b>2.3</b> Count normalization of Mov10 dataset</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="count-normalization.html"><a href="count-normalization.html#match-the-metadata-and-counts-data"><i class="fa fa-check"></i><b>2.3.1</b> 1. Match the metadata and counts data</a></li>
<li class="chapter" data-level="2.3.2" data-path="count-normalization.html"><a href="count-normalization.html#create-deseq2-object"><i class="fa fa-check"></i><b>2.3.2</b> 2. Create DESEq2 object</a></li>
<li class="chapter" data-level="2.3.3" data-path="count-normalization.html"><a href="count-normalization.html#generate-the-mov10-normalized-counts"><i class="fa fa-check"></i><b>2.3.3</b> 3. Generate the Mov10 normalized counts</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#sample-level-qc"><i class="fa fa-check"></i><b>3.1</b> Sample-level QC</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="quality-control.html"><a href="quality-control.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>3.1.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="3.1.2" data-path="quality-control.html"><a href="quality-control.html#hierarchical-clustering-heatmap"><i class="fa fa-check"></i><b>3.1.2</b> Hierarchical Clustering Heatmap</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#gene-level-qc"><i class="fa fa-check"></i><b>3.2</b> Gene-level QC</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#mov10-quality-assessment-and-exploratory-analysis-using-deseq2"><i class="fa fa-check"></i><b>3.3</b> Mov10 quality assessment and exploratory analysis using DESeq2</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="quality-control.html"><a href="quality-control.html#transform-normalized-counts-using-the-rlog-transformation"><i class="fa fa-check"></i><b>3.3.1</b> Transform normalized counts using the rlog transformation</a></li>
<li class="chapter" data-level="3.3.2" data-path="quality-control.html"><a href="quality-control.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>3.3.2</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="3.3.3" data-path="quality-control.html"><a href="quality-control.html#hierarchical-clustering"><i class="fa fa-check"></i><b>3.3.3</b> Hierarchical Clustering</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html"><i class="fa fa-check"></i><b>4</b> DGE analysis workflow</a>
<ul>
<li class="chapter" data-level="4.1" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#running-deseq2"><i class="fa fa-check"></i><b>4.1</b> Running DESeq2</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#design-formula"><i class="fa fa-check"></i><b>4.1.1</b> Design formula</a></li>
<li class="chapter" data-level="4.1.2" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#mov10-de-analysis"><i class="fa fa-check"></i><b>4.1.2</b> MOV10 DE analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#deseq2-differential-gene-expression-analysis-workflow"><i class="fa fa-check"></i><b>4.2</b> DESeq2 differential gene expression analysis workflow</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-1-estimate-size-factors"><i class="fa fa-check"></i><b>4.2.1</b> Step 1: Estimate size factors</a></li>
<li class="chapter" data-level="4.2.2" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-2-estimate-gene-wise-dispersion"><i class="fa fa-check"></i><b>4.2.2</b> Step 2: Estimate gene-wise dispersion</a></li>
<li class="chapter" data-level="4.2.3" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-3-fit-curve-to-gene-wise-dispersion-estimates"><i class="fa fa-check"></i><b>4.2.3</b> Step 3: Fit curve to gene-wise dispersion estimates</a></li>
<li class="chapter" data-level="4.2.4" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve"><i class="fa fa-check"></i><b>4.2.4</b> Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html"><i class="fa fa-check"></i><b>5</b> Model and hypothesis testing</a>
<ul>
<li class="chapter" data-level="5.1" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#generalized-linear-model-fit-for-each-gene"><i class="fa fa-check"></i><b>5.1</b> Generalized Linear Model fit for each gene</a></li>
<li class="chapter" data-level="5.2" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#shrunken-log2-foldchanges-lfc"><i class="fa fa-check"></i><b>5.2</b> Shrunken log2 foldchanges (LFC)</a></li>
<li class="chapter" data-level="5.3" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#statistical-test-for-lfc-estimates-wald-test"><i class="fa fa-check"></i><b>5.3</b> Statistical test for LFC estimates: Wald test</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#overall-deseq2-workflow"><i class="fa fa-check"></i><b>5.3.1</b> Overall DESeq2 workflow:</a></li>
<li class="chapter" data-level="5.3.2" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#creating-contrasts"><i class="fa fa-check"></i><b>5.3.2</b> Creating contrasts</a></li>
<li class="chapter" data-level="5.3.3" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#building-the-results-table"><i class="fa fa-check"></i><b>5.3.3</b> Building the results table</a></li>
<li class="chapter" data-level="5.3.4" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#ma-plot"><i class="fa fa-check"></i><b>5.3.4</b> MA Plot</a></li>
<li class="chapter" data-level="5.3.5" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#mov10-de-analysis-results-exploration"><i class="fa fa-check"></i><b>5.3.5</b> MOV10 DE analysis: results exploration</a></li>
<li class="chapter" data-level="5.3.6" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#multiple-test-correction"><i class="fa fa-check"></i><b>5.3.6</b> Multiple test correction</a></li>
<li class="chapter" data-level="5.3.7" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#mov10-de-analysis-control-versus-knockdown"><i class="fa fa-check"></i><b>5.3.7</b> MOV10 DE analysis: Control versus Knockdown</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#summarizing-results"><i class="fa fa-check"></i><b>5.4</b> Summarizing results</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#extracting-significant-differentially-expressed-genes"><i class="fa fa-check"></i><b>5.4.1</b> Extracting significant differentially expressed genes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html"><i class="fa fa-check"></i><b>6</b> Visualizing RNA-seq results</a>
<ul>
<li class="chapter" data-level="6.0.1" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html#plotting-signicant-de-genes"><i class="fa fa-check"></i><b>6.0.1</b> Plotting signicant DE genes</a></li>
<li class="chapter" data-level="6.0.2" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html#heatmap"><i class="fa fa-check"></i><b>6.0.2</b> Heatmap</a></li>
<li class="chapter" data-level="6.0.3" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html#volcano-plot"><i class="fa fa-check"></i><b>6.0.3</b> Volcano plot</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="summary-of-dge-workflow.html"><a href="summary-of-dge-workflow.html"><i class="fa fa-check"></i><b>7</b> Summary of DGE workflow</a></li>
<li class="chapter" data-level="8" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html"><i class="fa fa-check"></i><b>8</b> Functional analysis of RNAseq data</a>
<ul>
<li class="chapter" data-level="8.0.1" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#over-representation-analysis"><i class="fa fa-check"></i><b>8.0.1</b> Over-representation analysis</a></li>
<li class="chapter" data-level="8.0.2" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#clusterprofiler"><i class="fa fa-check"></i><b>8.0.2</b> clusterProfiler</a></li>
<li class="chapter" data-level="8.0.3" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#gene-set-enrichment-analysis-gsea"><i class="fa fa-check"></i><b>8.0.3</b> Gene set enrichment analysis (GSEA)</a></li>
<li class="chapter" data-level="8.0.4" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#other-tools-and-resources"><i class="fa fa-check"></i><b>8.0.4</b> Other tools and resources</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="codebook.html"><a href="codebook.html"><i class="fa fa-check"></i><b>9</b> Codebook</a>
<ul>
<li class="chapter" data-level="9.1" data-path="codebook.html"><a href="codebook.html#dge-analysis-overview"><i class="fa fa-check"></i><b>9.1</b> DGE analysis overview</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="codebook.html"><a href="codebook.html#review-of-the-dataset-1"><i class="fa fa-check"></i><b>9.1.1</b> Review of the dataset</a></li>
<li class="chapter" data-level="9.1.2" data-path="codebook.html"><a href="codebook.html#setting-up-1"><i class="fa fa-check"></i><b>9.1.2</b> Setting up</a></li>
<li class="chapter" data-level="9.1.3" data-path="codebook.html"><a href="codebook.html#dge-analysis-workflow-1"><i class="fa fa-check"></i><b>9.1.3</b> DGE analysis workflow</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="codebook.html"><a href="codebook.html#count-normalization-1"><i class="fa fa-check"></i><b>9.2</b> Count normalization</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="codebook.html"><a href="codebook.html#normalization-1"><i class="fa fa-check"></i><b>9.2.1</b> Normalization</a></li>
<li class="chapter" data-level="9.2.2" data-path="codebook.html"><a href="codebook.html#count-normalization-of-mov10-dataset-1"><i class="fa fa-check"></i><b>9.2.2</b> Count normalization of Mov10 dataset</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="codebook.html"><a href="codebook.html#dge-qc-analysis"><i class="fa fa-check"></i><b>9.3</b> DGE QC analysis</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="codebook.html"><a href="codebook.html#sample-level-qc-1"><i class="fa fa-check"></i><b>9.3.1</b> Sample-level QC</a></li>
<li class="chapter" data-level="9.3.2" data-path="codebook.html"><a href="codebook.html#mov10-quality-assessment-and-exploratory-analysis"><i class="fa fa-check"></i><b>9.3.2</b> Mov10 quality assessment and exploratory analysis</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="codebook.html"><a href="codebook.html#deg-analysis-workflow"><i class="fa fa-check"></i><b>9.4</b> DEG analysis workflow</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="codebook.html"><a href="codebook.html#running-deseq2-1"><i class="fa fa-check"></i><b>9.4.1</b> Running DESeq2</a></li>
<li class="chapter" data-level="9.4.2" data-path="codebook.html"><a href="codebook.html#deseq2-differential-gene-expression-analysis-workflow-1"><i class="fa fa-check"></i><b>9.4.2</b> DESeq2 differential gene expression analysis workflow</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="codebook.html"><a href="codebook.html#model-fitting"><i class="fa fa-check"></i><b>9.5</b> Model fitting</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="codebook.html"><a href="codebook.html#generalized-linear-model-fit-for-each-gene-1"><i class="fa fa-check"></i><b>9.5.1</b> Generalized Linear Model fit for each gene</a></li>
<li class="chapter" data-level="9.5.2" data-path="codebook.html"><a href="codebook.html#shrunken-log2-foldchanges-lfc-1"><i class="fa fa-check"></i><b>9.5.2</b> Shrunken log2 foldchanges (LFC)</a></li>
<li class="chapter" data-level="9.5.3" data-path="codebook.html"><a href="codebook.html#hypothesis-testing-using-the-wald-test"><i class="fa fa-check"></i><b>9.5.3</b> Hypothesis testing using the Wald test</a></li>
<li class="chapter" data-level="9.5.4" data-path="codebook.html"><a href="codebook.html#summarizing-results-1"><i class="fa fa-check"></i><b>9.5.4</b> Summarizing results</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="codebook.html"><a href="codebook.html#visualizing-rna-seq-results-1"><i class="fa fa-check"></i><b>9.6</b> Visualizing rna-seq results</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="codebook.html"><a href="codebook.html#plotting-signicant-de-genes-1"><i class="fa fa-check"></i><b>9.6.1</b> Plotting signicant DE genes</a></li>
<li class="chapter" data-level="9.6.2" data-path="codebook.html"><a href="codebook.html#heatmap-1"><i class="fa fa-check"></i><b>9.6.2</b> Heatmap</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="codebook.html"><a href="codebook.html#summary-of-differential-expression-analysis-workflow"><i class="fa fa-check"></i><b>9.7</b> Summary of differential expression analysis workflow</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="codebook.html"><a href="codebook.html#import-data-into-dds-object"><i class="fa fa-check"></i><b>9.7.1</b> 1. Import data into dds object:</a></li>
<li class="chapter" data-level="9.7.2" data-path="codebook.html"><a href="codebook.html#exploratory-data-analysis-pca-heirarchical-clustering---identifying-outliers-and-sources-of-variation-in-the-data"><i class="fa fa-check"></i><b>9.7.2</b> 2. Exploratory data analysis (PCA &amp; heirarchical clustering) - identifying outliers and sources of variation in the data:</a></li>
<li class="chapter" data-level="9.7.3" data-path="codebook.html"><a href="codebook.html#run-deseq2"><i class="fa fa-check"></i><b>9.7.3</b> 3. Run DESeq2:</a></li>
<li class="chapter" data-level="9.7.4" data-path="codebook.html"><a href="codebook.html#check-the-fit-of-the-dispersion-estimates"><i class="fa fa-check"></i><b>9.7.4</b> 4. Check the fit of the dispersion estimates:</a></li>
<li class="chapter" data-level="9.7.5" data-path="codebook.html"><a href="codebook.html#create-contrasts-to-perform-wald-testing-on-the-shrunken-log2-foldchanges-between-specific-conditions"><i class="fa fa-check"></i><b>9.7.5</b> 5. Create contrasts to perform Wald testing on the shrunken log2 foldchanges between specific conditions:</a></li>
<li class="chapter" data-level="9.7.6" data-path="codebook.html"><a href="codebook.html#output-significant-results"><i class="fa fa-check"></i><b>9.7.6</b> 6. Output significant results:</a></li>
<li class="chapter" data-level="9.7.7" data-path="codebook.html"><a href="codebook.html#visualize-results-volcano-plots-heatmaps-normalized-counts-plots-of-top-genes-etc."><i class="fa fa-check"></i><b>9.7.7</b> 7. Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc.</a></li>
<li class="chapter" data-level="9.7.8" data-path="codebook.html"><a href="codebook.html#make-sure-to-output-the-versions-of-all-tools-used-in-the-de-analysis"><i class="fa fa-check"></i><b>9.7.8</b> 8. Make sure to output the versions of all tools used in the DE analysis:</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="codebook.html"><a href="codebook.html#functional-analysis-of-rna-seq-data"><i class="fa fa-check"></i><b>9.8</b> Functional analysis of RNA-seq data</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="codebook.html"><a href="codebook.html#over-representation-analysis-1"><i class="fa fa-check"></i><b>9.8.1</b> Over-representation analysis</a></li>
<li class="chapter" data-level="9.8.2" data-path="codebook.html"><a href="codebook.html#clusterprofiler-1"><i class="fa fa-check"></i><b>9.8.2</b> clusterProfiler</a></li>
<li class="chapter" data-level="9.8.3" data-path="codebook.html"><a href="codebook.html#gene-set-enrichment-analysis-gsea-1"><i class="fa fa-check"></i><b>9.8.3</b> Gene set enrichment analysis (GSEA)</a></li>
<li class="chapter" data-level="9.8.4" data-path="codebook.html"><a href="codebook.html#other-tools-and-resources-1"><i class="fa fa-check"></i><b>9.8.4</b> Other tools and resources</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNA-seq-analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="count-normalization" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Count normalization<a href="count-normalization.html#count-normalization" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="normalization" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Normalization<a href="count-normalization.html#normalization" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The first step in the DE analysis workflow is count normalization, which is necessary to make accurate comparisons of gene expression between samples.</p>
<div class="float">
<img src="img/deseq_workflow_normalization_2018.png" width="400" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>The counts of mapped reads for each gene is proportional to the expression of RNA (“interesting”) in addition to many other factors (“uninteresting”). Normalization is the process of scaling raw count values to account for the “uninteresting” factors. In this way the expression levels are more comparable between and/or within samples.</p>
<p>The main factors often considered during normalization are:</p>
<ul>
<li><p><strong>Sequencing depth:</strong> Accounting for sequencing depth is necessary for comparison of gene expression between samples. In the example below, each gene appears to have doubled in expression in <em>Sample A</em> relative to <em>Sample B</em>, however this is a consequence of <em>Sample A</em> having double the sequencing depth.</p>
<div class="float">
<img src="img/normalization_methods_depth.png" width="400" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<blockquote>
<p><em><strong>NOTE:</strong> In the figure above, each pink and green rectangle represents a read aligned to a gene. Reads connected by dashed lines connect a read spanning an intron.</em></p>
</blockquote></li>
<li><p><strong>Gene length:</strong> Accounting for gene length is necessary for comparing expression between different genes within the same sample. In the example, <em>Gene X</em> and <em>Gene Y</em> have similar levels of expression, but the number of reads mapped to <em>Gene X</em> would be many more than the number mapped to <em>Gene Y</em> because <em>Gene X</em> is longer.</p>
<div class="float">
<img src="img/normalization_methods_length.png" width="200" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div></li>
<li><p><strong>RNA composition:</strong> A few highly differentially expressed genes between samples, differences in the number of genes expressed between samples, or presence of contamination can skew some types of normalization methods. Accounting for RNA composition is recommended for accurate comparison of expression between samples, and is particularly important when performing differential expression analyses [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">1</a>].</p>
<p>In the example, if we were to divide each sample by the total number of counts to normalize, the counts would be greatly skewed by the DE gene, which takes up most of the counts for <em>Sample A</em>, but not <em>Sample B</em>. Most other genes for <em>Sample A</em> would be divided by the larger number of total counts and appear to be less expressed than those same genes in <em>Sample B</em>.<br />
<img src="img/normalization_methods_composition.png" width="400" alt="Alt text" /></p></li>
</ul>
<p><strong><em>While normalization is essential for differential expression analyses, it is also necessary for exploratory data analysis, visualization of data, and whenever you are exploring or comparing counts between or within samples.</em></strong></p>
<div id="common-normalization-methods" class="section level3 hasAnchor" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> Common normalization methods<a href="count-normalization.html#common-normalization-methods" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Several common normalization methods exist to account for these differences:</p>
<table>
<colgroup>
<col width="25%" />
<col width="25%" />
<col width="25%" />
<col width="25%" />
</colgroup>
<thead>
<tr class="header">
<th>Normalization method</th>
<th>Description</th>
<th>Accounted factors</th>
<th>Recommendations for use</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>TPM</strong> (transcripts per kilobase million)</td>
<td>counts per length of transcript (kb) per million reads mapped</td>
<td>sequencing depth and gene length</td>
<td>gene count comparisons within a sample or between samples of the same sample group; <strong>NOT for DE analysis</strong></td>
</tr>
<tr class="even">
<td><strong>RPKM/FPKM</strong> (reads/fragments per kilobase of exon per million reads/fragments mapped)</td>
<td>similar to TPM</td>
<td>sequencing depth and gene length</td>
<td>gene count comparisons between genes within a sample; <strong>NOT for between sample comparisons or DE analysis</strong></td>
</tr>
<tr class="odd">
<td>DESeq2’s <strong>median of ratios</strong> [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">1</a>]</td>
<td>counts divided by sample-specific size factors determined by median ratio of gene counts relative to geometric mean per gene</td>
<td>sequencing depth and RNA composition</td>
<td>gene count comparisons between samples and for <strong>DE analysis</strong>; <strong>NOT for within sample comparisons</strong></td>
</tr>
<tr class="even">
<td>EdgeR’s <strong>trimmed mean of M values (TMM)</strong> [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">2</a>]</td>
<td>uses a weighted trimmed mean of the log expression ratios between samples</td>
<td>sequencing depth, RNA composition</td>
<td>gene count comparisons between samples and for <strong>DE analysis</strong>; <strong>NOT for within sample comparisons</strong></td>
</tr>
</tbody>
</table>
</div>
<div id="rpkmfpkm-not-recommended" class="section level3 hasAnchor" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> RPKM/FPKM (not recommended)<a href="count-normalization.html#rpkmfpkm-not-recommended" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>While TPM and RPKM/FPKM normalization methods both account for sequencing depth and gene length, RPKM/FPKM are not recommended. <strong>The reason is that the normalized count values output by the RPKM/FPKM method are not comparable between samples.</strong></p>
<p>Using RPKM/FPKM normalization, the total number of RPKM/FPKM normalized counts for each sample will be different. Therefore, you cannot compare the normalized counts for each gene equally between samples.</p>
<p><strong>RPKM-normalized counts table</strong></p>
<table>
<thead>
<tr class="header">
<th>gene</th>
<th align="center">sampleA</th>
<th align="center">sampleB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>XCR1</td>
<td align="center">5.5</td>
<td align="center">5.5</td>
</tr>
<tr class="even">
<td>WASHC1</td>
<td align="center">73.4</td>
<td align="center">21.8</td>
</tr>
<tr class="odd">
<td>…</td>
<td align="center">…</td>
<td align="center">…</td>
</tr>
<tr class="even">
<td>Total RPKM-normalized counts</td>
<td align="center">1,000,000</td>
<td align="center">1,500,000</td>
</tr>
</tbody>
</table>
<p>For example, in the table above, SampleA has a greater proportion of counts associated with XCR1 (5.5/1,000,000) than does sampleB (5.5/1,500,000) even though the RPKM count values are the same. Therefore, we cannot directly compare the counts for XCR1 (or any other gene) between sampleA and sampleB because the total number of normalized counts are different between samples.</p>
</div>
<div id="deseq2-normalized-counts-median-of-ratios-method" class="section level3 hasAnchor" number="2.1.3">
<h3><span class="header-section-number">2.1.3</span> DESeq2-normalized counts: Median of ratios method<a href="count-normalization.html#deseq2-normalized-counts-median-of-ratios-method" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Since tools for differential expression analysis are comparing the counts between sample groups for the same gene, gene length does not need to be accounted for by the tool. However, <strong>sequencing depth</strong> and <strong>RNA composition</strong> do need to be taken into account.</p>
<p>To normalize for sequencing depth and RNA composition, DESeq2 uses the median of ratios method. On the user-end there is only one step, but on the back-end there are multiple steps involved, as described below.</p>
<blockquote>
<p><strong>NOTE:</strong> The steps below describe in detail some of the steps performed by DESeq2 when you run a single function to get DE genes. Basically, for a typical RNA-seq analysis, <strong>you would not run these steps individually</strong>.</p>
</blockquote>
<p><strong>Step 1: creates a pseudo-reference sample (row-wise geometric mean)</strong></p>
</div>
</div>
<div id="geometric-mean" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Geometric Mean<a href="count-normalization.html#geometric-mean" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The <strong>geometric mean</strong> of a set of <span class="math inline">\(n\)</span> positive numbers <span class="math inline">\(x_1, x_2, \dots, x_n\)</span> is defined as the <span class="math inline">\(n\)</span>th root of the product of the numbers. It is mathematically expressed as:</p>
<p><span class="math display">\[
\text{Geometric Mean} = \left( \prod_{i=1}^{n} x_i \right)^{\frac{1}{n}}
\]</span></p>
<p>where:
- <span class="math inline">\(x_1, x_2, \dots, x_n\)</span> are the individual numbers in the set,
- <span class="math inline">\(\prod_{i=1}^{n} x_i\)</span> denotes the product of all the numbers.</p>
<p>For each gene, a pseudo-reference sample is created that is equal to the geometric mean across all samples. The geometric mean is less affected by extremely high or low values than the arithmetic mean, making it more appropriate for skewed data, such as count data that has a large number of genes with low counts.</p>
<table>
<thead>
<tr class="header">
<th>gene</th>
<th align="center">sampleA</th>
<th align="center">sampleB</th>
<th align="center">pseudo-reference sample</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EF2A</td>
<td align="center">1489</td>
<td align="center">906</td>
<td align="center">sqrt(1489 * 906) = <strong>1161.5</strong></td>
</tr>
<tr class="even">
<td>ABCD1</td>
<td align="center">22</td>
<td align="center">13</td>
<td align="center">sqrt(22 * 13) = <strong>17.7</strong></td>
</tr>
<tr class="odd">
<td>…</td>
<td align="center">…</td>
<td align="center">…</td>
<td align="center">…</td>
</tr>
</tbody>
</table>
<p><strong>Step 2: calculates ratio of each sample to the reference</strong></p>
<p>For every gene in a sample, the ratios (sample/ref) are calculated (as shown below). This is performed for each sample in the dataset. Since the majority of genes are not differentially expressed, the majority of genes in each sample should have similar ratios within the sample.</p>
<table>
<colgroup>
<col width="12%" />
<col width="17%" />
<col width="17%" />
<col width="17%" />
<col width="17%" />
<col width="17%" />
</colgroup>
<thead>
<tr class="header">
<th>gene</th>
<th align="center">sampleA</th>
<th align="center">sampleB</th>
<th align="center">pseudo-reference sample</th>
<th align="center">ratio of sampleA/ref</th>
<th align="center">ratio of sampleB/ref</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EF2A</td>
<td align="center">1489</td>
<td align="center">906</td>
<td align="center">1161.5</td>
<td align="center">1489/1161.5 = <strong>1.28</strong></td>
<td align="center">906/1161.5 = <strong>0.78</strong></td>
</tr>
<tr class="even">
<td>ABCD1</td>
<td align="center">22</td>
<td align="center">13</td>
<td align="center">16.9</td>
<td align="center">22/16.9 = <strong>1.30</strong></td>
<td align="center">13/16.9 = <strong>0.77</strong></td>
</tr>
<tr class="odd">
<td>MEFV</td>
<td align="center">793</td>
<td align="center">410</td>
<td align="center">570.2</td>
<td align="center">793/570.2 = <strong>1.39</strong></td>
<td align="center">410/570.2 = <strong>0.72</strong></td>
</tr>
<tr class="even">
<td>BAG1</td>
<td align="center">76</td>
<td align="center">42</td>
<td align="center">56.5</td>
<td align="center">76/56.5 = <strong>1.35</strong></td>
<td align="center">42/56.5 = <strong>0.74</strong></td>
</tr>
<tr class="odd">
<td>MOV10</td>
<td align="center">521</td>
<td align="center">1196</td>
<td align="center">883.7</td>
<td align="center">521/883.7 = <strong>0.590</strong></td>
<td align="center">1196/883.7 = <strong>1.35</strong></td>
</tr>
<tr class="even">
<td>…</td>
<td align="center">…</td>
<td align="center">…</td>
<td align="center">…</td>
<td align="center"></td>
<td align="center"></td>
</tr>
</tbody>
</table>
<p><strong>Step 3: calculate the normalization factor for each sample (size factor)</strong></p>
<p>The median value (column-wise for the above table) of all ratios for a given sample is taken as the normalization factor (size factor) for that sample, as calculated below. Notice that the differentially expressed genes should not affect the median value:</p>
<p><code>normalization_factor_sampleA &lt;- median(c(1.28, 1.3, 1.39, 1.35, 0.59))</code></p>
<p><code>normalization_factor_sampleB &lt;- median(c(0.78, 0.77, 0.72, 0.74, 1.35))</code></p>
<p>The figure below illustrates the median value for the distribution of all gene ratios for a single sample (frequency is on the y-axis).</p>
<div class="float">
<img src="img/deseq_median_of_ratios.png" width="400" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>The median of ratios method makes the assumption that not <strong>ALL</strong> genes are differentially expressed; therefore, the normalization factors should account for sequencing depth and RNA composition of the sample (large outlier genes will not represent the median ratio values). <strong>This method is robust to imbalance in up-/down-regulation and large numbers of differentially expressed genes.</strong></p>
<blockquote>
<p>Usually these size factors are around 1, if you see large variations between samples it is important to take note since it might indicate the presence of extreme outliers.</p>
</blockquote>
<p><strong>Step 4: calculate the normalized count values using the normalization factor</strong></p>
<p>This is performed by dividing each raw count value in a given sample by that sample’s normalization factor to generate normalized count values. This is performed for all count values (every gene in every sample). For example, if the median ratio for SampleA was 1.3 and the median ratio for SampleB was 0.77, you could calculate normalized counts as follows:</p>
<p>SampleA median ratio = 1.3</p>
<p>SampleB median ratio = 0.77</p>
<p><strong>Raw Counts</strong></p>
<table>
<thead>
<tr class="header">
<th>gene</th>
<th align="center">sampleA</th>
<th align="center">sampleB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EF2A</td>
<td align="center">1489</td>
<td align="center">906</td>
</tr>
<tr class="even">
<td>ABCD1</td>
<td align="center">22</td>
<td align="center">13</td>
</tr>
<tr class="odd">
<td>…</td>
<td align="center">…</td>
<td align="center">…</td>
</tr>
</tbody>
</table>
<p><strong>Normalized Counts</strong></p>
<table>
<thead>
<tr class="header">
<th>gene</th>
<th align="center">sampleA</th>
<th align="center">sampleB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>EF2A</td>
<td align="center">1489 / 1.3 = <strong>1145.39</strong></td>
<td align="center">906 / 0.77 = <strong>1176.62</strong></td>
</tr>
<tr class="even">
<td>ABCD1</td>
<td align="center">22 / 1.3 = <strong>16.92</strong></td>
<td align="center">13 / 0.77 = <strong>16.88</strong></td>
</tr>
<tr class="odd">
<td>…</td>
<td align="center">…</td>
<td align="center">…</td>
</tr>
</tbody>
</table>
<blockquote>
<p>Please note that normalized count values are not whole numbers.</p>
</blockquote>
<hr />
<p><strong>Exercise</strong></p>
<p>Determine the normalized counts for your gene of interest, PD1, given the raw counts and size factors below.</p>
<p>NOTE: You will need to run the code below to generate the raw counts dataframe (PD1) and the size factor vector (size_factors), then use these objects to determine the normalized counts values:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="count-normalization.html#cb9-1" tabindex="-1"></a></span>
<span id="cb9-2"><a href="count-normalization.html#cb9-2" tabindex="-1"></a><span class="co"># Raw counts for PD1</span></span>
<span id="cb9-3"><a href="count-normalization.html#cb9-3" tabindex="-1"></a>PD1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">58</span>, <span class="dv">17</span>, <span class="dv">97</span>, <span class="dv">83</span>, <span class="dv">10</span>)</span>
<span id="cb9-4"><a href="count-normalization.html#cb9-4" tabindex="-1"></a><span class="fu">names</span>(PD1) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;Sample&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb9-5"><a href="count-normalization.html#cb9-5" tabindex="-1"></a>PD1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(PD1)</span>
<span id="cb9-6"><a href="count-normalization.html#cb9-6" tabindex="-1"></a>PD1 <span class="ot">&lt;-</span> <span class="fu">t</span>(PD1)</span>
<span id="cb9-7"><a href="count-normalization.html#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="count-normalization.html#cb9-8" tabindex="-1"></a><span class="co"># Size factors for each sample</span></span>
<span id="cb9-9"><a href="count-normalization.html#cb9-9" tabindex="-1"></a>size_factors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">1.32</span>, <span class="fl">0.70</span>, <span class="fl">1.04</span>, <span class="fl">1.27</span>, <span class="fl">1.11</span>, <span class="fl">0.85</span>)</span></code></pre></div>
<hr />
</div>
<div id="count-normalization-of-mov10-dataset" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Count normalization of Mov10 dataset<a href="count-normalization.html#count-normalization-of-mov10-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that we know the theory of count normalization, we will normalize the counts for the Mov10 dataset using DESeq2. This requires a few steps:</p>
<ol style="list-style-type: decimal">
<li>Ensure the row names of the metadata dataframe are present and in the same order as the column names of the counts dataframe.</li>
<li>Create a <code>DESeqDataSet</code> object</li>
<li>Generate the normalized counts</li>
</ol>
<div id="match-the-metadata-and-counts-data" class="section level3 hasAnchor" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> 1. Match the metadata and counts data<a href="count-normalization.html#match-the-metadata-and-counts-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We should always make sure that we have sample names that match between the two files, and that the samples are in the right order. DESeq2 will output an error if this is not the case.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="count-normalization.html#cb10-1" tabindex="-1"></a><span class="do">### Check that sample names match in both files</span></span>
<span id="cb10-2"><a href="count-normalization.html#cb10-2" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">colnames</span>(data) <span class="sc">%in%</span> <span class="fu">rownames</span>(meta))</span>
<span id="cb10-3"><a href="count-normalization.html#cb10-3" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">colnames</span>(data) <span class="sc">==</span> <span class="fu">rownames</span>(meta))</span></code></pre></div>
<p>The colnames of our data don’t match the rownames of our metadata so we need to reorder them. We can use the <code>match</code> function:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="count-normalization.html#cb11-1" tabindex="-1"></a><span class="co"># write the code to match the colnames of data to the rownames of meta</span></span>
<span id="cb11-2"><a href="count-normalization.html#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="count-normalization.html#cb11-3" tabindex="-1"></a><span class="co"># check that the rownames of the metadata match the colnames of the data</span></span></code></pre></div>
<hr />
</div>
<div id="create-deseq2-object" class="section level3 hasAnchor" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> 2. Create DESEq2 object<a href="count-normalization.html#create-deseq2-object" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Bioconductor software packages often define and use a custom class within R for storing data (input data, intermediate data and also result data). These custom data structures are similar to <code>lists</code> in that they can contain multiple different data types/structures within them. But, unlike lists they have pre-specified <code>data slots</code>, which hold specific types/classes of data. The data stored in these pre-specified slots can be accessed by using specific package-defined functions.</p>
<p>Let’s start by creating the <code>DESeqDataSet</code> object and then we can talk a bit more about what is stored inside it. To create the object we will need the <strong>count matrix</strong> and the <strong>metadata</strong> table as input. We will also need to specify a <strong>design formula</strong>. The design formula specifies the column(s) in the metadata table and how they should be used in the analysis. For our dataset we only have one column we are interested in, that is <code>~sampletype</code>. This column has three factor levels, which tells DESeq2 that for each gene we want to evaluate gene expression change with respect to these different levels.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="count-normalization.html#cb12-1" tabindex="-1"></a><span class="do">## Create DESeq2Dataset object</span></span>
<span id="cb12-2"><a href="count-normalization.html#cb12-2" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> data, <span class="at">colData =</span> meta, <span class="at">design =</span> <span class="sc">~</span> sampletype)</span></code></pre></div>
<div class="float">
<img src="img/deseq_obj1.png" alt="deseq1" />
<div class="figcaption">deseq1</div>
</div>
<p>You can use DESeq-specific functions to access the different slots and retrieve information, if you wish. For example, suppose we wanted the original count matrix we would use <code>counts()</code>:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="count-normalization.html#cb13-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">counts</span>(dds[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]))</span></code></pre></div>
<p>As we go through the workflow we will use the relevant functions to check what information gets stored inside our object. We can also run:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="count-normalization.html#cb14-1" tabindex="-1"></a><span class="fu">slotNames</span>(dds)</span></code></pre></div>
</div>
<div id="generate-the-mov10-normalized-counts" class="section level3 hasAnchor" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> 3. Generate the Mov10 normalized counts<a href="count-normalization.html#generate-the-mov10-normalized-counts" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The next step is to normalize the count data in order to be able to make fair gene comparisons between samples.</p>
<div class="float">
<img src="img/deseq_workflow_normalization_2018.png" width="400" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>To perform the <strong>median of ratios method</strong> of normalization, DESeq2 has a single <code>estimateSizeFactors()</code> function that will generate size factors for us. We will use the function in the example below, but <strong>in a typical RNA-seq analysis this step is automatically performed by the <code>DESeq()</code> function</strong>, which we will see later.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="count-normalization.html#cb15-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span></code></pre></div>
<p>By assigning the results back to the <code>dds</code> object we are filling in the slots of the <code>DESeqDataSet</code> object with the appropriate information. We can take a look at the normalization factor applied to each sample using:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="count-normalization.html#cb16-1" tabindex="-1"></a><span class="fu">sizeFactors</span>(dds)</span></code></pre></div>
<p>Now, to retrieve the normalized counts matrix from <code>dds</code>, we use the <code>counts()</code> function and add the argument <code>normalized=TRUE</code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="count-normalization.html#cb17-1" tabindex="-1"></a>normalized_counts <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds, <span class="at">normalized=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>We can save this normalized data matrix to file for later use:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="count-normalization.html#cb18-1" tabindex="-1"></a><span class="fu">write.table</span>(normalized_counts, <span class="at">file=</span><span class="st">&quot;data/normalized_counts.txt&quot;</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote=</span>F, <span class="at">col.names=</span><span class="cn">NA</span>)</span></code></pre></div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differential-gene-expression-dge-analysis-overview.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="quality-control.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["RNA-seq-analysis.pdf", "RNA-seq-analysis.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
