<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 DGE analysis workflow | RNA-seq-analysis</title>
  <meta name="description" content="RNA-seq analysis workshop" />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 DGE analysis workflow | RNA-seq-analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bookdown.org/ggiaever/RNA-seq-analysis/img/football22.png" />
  <meta property="og:description" content="RNA-seq analysis workshop" />
  <meta name="github-repo" content="gurinina/RNA-seq-analysis" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 DGE analysis workflow | RNA-seq-analysis" />
  
  <meta name="twitter:description" content="RNA-seq analysis workshop" />
  <meta name="twitter:image" content="https://bookdown.org/ggiaever/RNA-seq-analysis/img/football22.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="quality-control.html"/>
<link rel="next" href="model-and-hypothesis-testing.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/vis-9.1.0/vis-network.min.css" rel="stylesheet" />
<script src="libs/vis-9.1.0/vis-network.min.js"></script>
<script src="libs/visNetwork-binding-2.1.2/visNetwork.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="css/git.css" type="text/css" />
<link rel="stylesheet" href="css/toc.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>RNA-seq analysis workshop</a></li>
<li class="chapter" data-level="1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html"><i class="fa fa-check"></i><b>1</b> Differential gene expression (DGE) analysis overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#review-of-the-dataset"><i class="fa fa-check"></i><b>1.1</b> Review of the dataset</a></li>
<li class="chapter" data-level="1.2" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#setting-up"><i class="fa fa-check"></i><b>1.2</b> Setting up</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#loading-libraries"><i class="fa fa-check"></i><b>1.2.1</b> Loading libraries</a></li>
<li class="chapter" data-level="1.2.2" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#loading-data"><i class="fa fa-check"></i><b>1.2.2</b> Loading data</a></li>
<li class="chapter" data-level="1.2.3" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#viewing-data"><i class="fa fa-check"></i><b>1.2.3</b> Viewing data</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#differential-gene-expression-analysis-overview"><i class="fa fa-check"></i><b>1.3</b> Differential gene expression analysis overview</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#rna-seq-count-distribution"><i class="fa fa-check"></i><b>1.3.1</b> RNA-seq count distribution</a></li>
<li class="chapter" data-level="1.3.2" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#modeling-count-data"><i class="fa fa-check"></i><b>1.3.2</b> Modeling count data</a></li>
<li class="chapter" data-level="1.3.3" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#improving-mean-estimates-i.e.-reducing-variance-with-biological-replicates"><i class="fa fa-check"></i><b>1.3.3</b> Improving mean estimates (i.e. reducing variance) with biological replicates</a></li>
<li class="chapter" data-level="1.3.4" data-path="differential-gene-expression-dge-analysis-overview.html"><a href="differential-gene-expression-dge-analysis-overview.html#differential-expression-analysis-workflow"><i class="fa fa-check"></i><b>1.3.4</b> Differential expression analysis workflow</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="count-normalization.html"><a href="count-normalization.html"><i class="fa fa-check"></i><b>2</b> Count normalization</a>
<ul>
<li class="chapter" data-level="2.1" data-path="count-normalization.html"><a href="count-normalization.html#normalization"><i class="fa fa-check"></i><b>2.1</b> Normalization</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="count-normalization.html"><a href="count-normalization.html#common-normalization-methods"><i class="fa fa-check"></i><b>2.1.1</b> Common normalization methods</a></li>
<li class="chapter" data-level="2.1.2" data-path="count-normalization.html"><a href="count-normalization.html#rpkmfpkm-not-recommended"><i class="fa fa-check"></i><b>2.1.2</b> RPKM/FPKM (not recommended)</a></li>
<li class="chapter" data-level="2.1.3" data-path="count-normalization.html"><a href="count-normalization.html#deseq2-normalized-counts-median-of-ratios-method"><i class="fa fa-check"></i><b>2.1.3</b> DESeq2-normalized counts: Median of ratios method</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="count-normalization.html"><a href="count-normalization.html#geometric-mean"><i class="fa fa-check"></i><b>2.2</b> Geometric Mean</a></li>
<li class="chapter" data-level="2.3" data-path="count-normalization.html"><a href="count-normalization.html#count-normalization-of-mov10-dataset"><i class="fa fa-check"></i><b>2.3</b> Count normalization of Mov10 dataset</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="count-normalization.html"><a href="count-normalization.html#match-the-metadata-and-counts-data"><i class="fa fa-check"></i><b>2.3.1</b> 1. Match the metadata and counts data</a></li>
<li class="chapter" data-level="2.3.2" data-path="count-normalization.html"><a href="count-normalization.html#create-deseq2-object"><i class="fa fa-check"></i><b>2.3.2</b> 2. Create DESEq2 object</a></li>
<li class="chapter" data-level="2.3.3" data-path="count-normalization.html"><a href="count-normalization.html#generate-the-mov10-normalized-counts"><i class="fa fa-check"></i><b>2.3.3</b> 3. Generate the Mov10 normalized counts</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#sample-level-qc"><i class="fa fa-check"></i><b>3.1</b> Sample-level QC</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="quality-control.html"><a href="quality-control.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>3.1.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="3.1.2" data-path="quality-control.html"><a href="quality-control.html#hierarchical-clustering-heatmap"><i class="fa fa-check"></i><b>3.1.2</b> Hierarchical Clustering Heatmap</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#gene-level-qc"><i class="fa fa-check"></i><b>3.2</b> Gene-level QC</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#mov10-quality-assessment-and-exploratory-analysis-using-deseq2"><i class="fa fa-check"></i><b>3.3</b> Mov10 quality assessment and exploratory analysis using DESeq2</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="quality-control.html"><a href="quality-control.html#transform-normalized-counts-using-the-rlog-transformation"><i class="fa fa-check"></i><b>3.3.1</b> Transform normalized counts using the rlog transformation</a></li>
<li class="chapter" data-level="3.3.2" data-path="quality-control.html"><a href="quality-control.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>3.3.2</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="3.3.3" data-path="quality-control.html"><a href="quality-control.html#hierarchical-clustering"><i class="fa fa-check"></i><b>3.3.3</b> Hierarchical Clustering</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html"><i class="fa fa-check"></i><b>4</b> DGE analysis workflow</a>
<ul>
<li class="chapter" data-level="4.1" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#running-deseq2"><i class="fa fa-check"></i><b>4.1</b> Running DESeq2</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#design-formula"><i class="fa fa-check"></i><b>4.1.1</b> Design formula</a></li>
<li class="chapter" data-level="4.1.2" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#mov10-de-analysis"><i class="fa fa-check"></i><b>4.1.2</b> MOV10 DE analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#deseq2-differential-gene-expression-analysis-workflow"><i class="fa fa-check"></i><b>4.2</b> DESeq2 differential gene expression analysis workflow</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-1-estimate-size-factors"><i class="fa fa-check"></i><b>4.2.1</b> Step 1: Estimate size factors</a></li>
<li class="chapter" data-level="4.2.2" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-2-estimate-gene-wise-dispersion"><i class="fa fa-check"></i><b>4.2.2</b> Step 2: Estimate gene-wise dispersion</a></li>
<li class="chapter" data-level="4.2.3" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-3-fit-curve-to-gene-wise-dispersion-estimates"><i class="fa fa-check"></i><b>4.2.3</b> Step 3: Fit curve to gene-wise dispersion estimates</a></li>
<li class="chapter" data-level="4.2.4" data-path="dge-analysis-workflow.html"><a href="dge-analysis-workflow.html#step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve"><i class="fa fa-check"></i><b>4.2.4</b> Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html"><i class="fa fa-check"></i><b>5</b> Model and hypothesis testing</a>
<ul>
<li class="chapter" data-level="5.1" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#generalized-linear-model-fit-for-each-gene"><i class="fa fa-check"></i><b>5.1</b> Generalized Linear Model fit for each gene</a></li>
<li class="chapter" data-level="5.2" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#shrunken-log2-foldchanges-lfc"><i class="fa fa-check"></i><b>5.2</b> Shrunken log2 foldchanges (LFC)</a></li>
<li class="chapter" data-level="5.3" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#statistical-test-for-lfc-estimates-wald-test"><i class="fa fa-check"></i><b>5.3</b> Statistical test for LFC estimates: Wald test</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#overall-deseq2-workflow"><i class="fa fa-check"></i><b>5.3.1</b> Overall DESeq2 workflow:</a></li>
<li class="chapter" data-level="5.3.2" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#creating-contrasts"><i class="fa fa-check"></i><b>5.3.2</b> Creating contrasts</a></li>
<li class="chapter" data-level="5.3.3" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#building-the-results-table"><i class="fa fa-check"></i><b>5.3.3</b> Building the results table</a></li>
<li class="chapter" data-level="5.3.4" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#ma-plot"><i class="fa fa-check"></i><b>5.3.4</b> MA Plot</a></li>
<li class="chapter" data-level="5.3.5" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#mov10-de-analysis-results-exploration"><i class="fa fa-check"></i><b>5.3.5</b> MOV10 DE analysis: results exploration</a></li>
<li class="chapter" data-level="5.3.6" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#multiple-test-correction"><i class="fa fa-check"></i><b>5.3.6</b> Multiple test correction</a></li>
<li class="chapter" data-level="5.3.7" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#mov10-de-analysis-control-versus-knockdown"><i class="fa fa-check"></i><b>5.3.7</b> MOV10 DE analysis: Control versus Knockdown</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#summarizing-results"><i class="fa fa-check"></i><b>5.4</b> Summarizing results</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="model-and-hypothesis-testing.html"><a href="model-and-hypothesis-testing.html#extracting-significant-differentially-expressed-genes"><i class="fa fa-check"></i><b>5.4.1</b> Extracting significant differentially expressed genes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html"><i class="fa fa-check"></i><b>6</b> Visualizing RNA-seq results</a>
<ul>
<li class="chapter" data-level="6.0.1" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html#plotting-signicant-de-genes"><i class="fa fa-check"></i><b>6.0.1</b> Plotting signicant DE genes</a></li>
<li class="chapter" data-level="6.0.2" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html#heatmap"><i class="fa fa-check"></i><b>6.0.2</b> Heatmap</a></li>
<li class="chapter" data-level="6.0.3" data-path="visualizing-rna-seq-results.html"><a href="visualizing-rna-seq-results.html#volcano-plot"><i class="fa fa-check"></i><b>6.0.3</b> Volcano plot</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="summary-of-dge-workflow.html"><a href="summary-of-dge-workflow.html"><i class="fa fa-check"></i><b>7</b> Summary of DGE workflow</a></li>
<li class="chapter" data-level="8" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html"><i class="fa fa-check"></i><b>8</b> Functional analysis of RNAseq data</a>
<ul>
<li class="chapter" data-level="8.0.1" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#over-representation-analysis"><i class="fa fa-check"></i><b>8.0.1</b> Over-representation analysis</a></li>
<li class="chapter" data-level="8.0.2" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#clusterprofiler"><i class="fa fa-check"></i><b>8.0.2</b> clusterProfiler</a></li>
<li class="chapter" data-level="8.0.3" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#gene-set-enrichment-analysis-gsea"><i class="fa fa-check"></i><b>8.0.3</b> Gene set enrichment analysis (GSEA)</a></li>
<li class="chapter" data-level="8.0.4" data-path="functional-analysis-of-rnaseq-data.html"><a href="functional-analysis-of-rnaseq-data.html#other-tools-and-resources"><i class="fa fa-check"></i><b>8.0.4</b> Other tools and resources</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="codebook.html"><a href="codebook.html"><i class="fa fa-check"></i><b>9</b> Codebook</a>
<ul>
<li class="chapter" data-level="9.1" data-path="codebook.html"><a href="codebook.html#dge-analysis-overview"><i class="fa fa-check"></i><b>9.1</b> DGE analysis overview</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="codebook.html"><a href="codebook.html#review-of-the-dataset-1"><i class="fa fa-check"></i><b>9.1.1</b> Review of the dataset</a></li>
<li class="chapter" data-level="9.1.2" data-path="codebook.html"><a href="codebook.html#setting-up-1"><i class="fa fa-check"></i><b>9.1.2</b> Setting up</a></li>
<li class="chapter" data-level="9.1.3" data-path="codebook.html"><a href="codebook.html#dge-analysis-workflow-1"><i class="fa fa-check"></i><b>9.1.3</b> DGE analysis workflow</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="codebook.html"><a href="codebook.html#count-normalization-1"><i class="fa fa-check"></i><b>9.2</b> Count normalization</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="codebook.html"><a href="codebook.html#normalization-1"><i class="fa fa-check"></i><b>9.2.1</b> Normalization</a></li>
<li class="chapter" data-level="9.2.2" data-path="codebook.html"><a href="codebook.html#count-normalization-of-mov10-dataset-1"><i class="fa fa-check"></i><b>9.2.2</b> Count normalization of Mov10 dataset</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="codebook.html"><a href="codebook.html#dge-qc-analysis"><i class="fa fa-check"></i><b>9.3</b> DGE QC analysis</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="codebook.html"><a href="codebook.html#sample-level-qc-1"><i class="fa fa-check"></i><b>9.3.1</b> Sample-level QC</a></li>
<li class="chapter" data-level="9.3.2" data-path="codebook.html"><a href="codebook.html#mov10-quality-assessment-and-exploratory-analysis"><i class="fa fa-check"></i><b>9.3.2</b> Mov10 quality assessment and exploratory analysis</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="codebook.html"><a href="codebook.html#deg-analysis-workflow"><i class="fa fa-check"></i><b>9.4</b> DEG analysis workflow</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="codebook.html"><a href="codebook.html#running-deseq2-1"><i class="fa fa-check"></i><b>9.4.1</b> Running DESeq2</a></li>
<li class="chapter" data-level="9.4.2" data-path="codebook.html"><a href="codebook.html#deseq2-differential-gene-expression-analysis-workflow-1"><i class="fa fa-check"></i><b>9.4.2</b> DESeq2 differential gene expression analysis workflow</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="codebook.html"><a href="codebook.html#model-fitting"><i class="fa fa-check"></i><b>9.5</b> Model fitting</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="codebook.html"><a href="codebook.html#generalized-linear-model-fit-for-each-gene-1"><i class="fa fa-check"></i><b>9.5.1</b> Generalized Linear Model fit for each gene</a></li>
<li class="chapter" data-level="9.5.2" data-path="codebook.html"><a href="codebook.html#shrunken-log2-foldchanges-lfc-1"><i class="fa fa-check"></i><b>9.5.2</b> Shrunken log2 foldchanges (LFC)</a></li>
<li class="chapter" data-level="9.5.3" data-path="codebook.html"><a href="codebook.html#hypothesis-testing-using-the-wald-test"><i class="fa fa-check"></i><b>9.5.3</b> Hypothesis testing using the Wald test</a></li>
<li class="chapter" data-level="9.5.4" data-path="codebook.html"><a href="codebook.html#summarizing-results-1"><i class="fa fa-check"></i><b>9.5.4</b> Summarizing results</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="codebook.html"><a href="codebook.html#visualizing-rna-seq-results-1"><i class="fa fa-check"></i><b>9.6</b> Visualizing rna-seq results</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="codebook.html"><a href="codebook.html#plotting-signicant-de-genes-1"><i class="fa fa-check"></i><b>9.6.1</b> Plotting signicant DE genes</a></li>
<li class="chapter" data-level="9.6.2" data-path="codebook.html"><a href="codebook.html#heatmap-1"><i class="fa fa-check"></i><b>9.6.2</b> Heatmap</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="codebook.html"><a href="codebook.html#summary-of-differential-expression-analysis-workflow"><i class="fa fa-check"></i><b>9.7</b> Summary of differential expression analysis workflow</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="codebook.html"><a href="codebook.html#import-data-into-dds-object"><i class="fa fa-check"></i><b>9.7.1</b> 1. Import data into dds object:</a></li>
<li class="chapter" data-level="9.7.2" data-path="codebook.html"><a href="codebook.html#exploratory-data-analysis-pca-heirarchical-clustering---identifying-outliers-and-sources-of-variation-in-the-data"><i class="fa fa-check"></i><b>9.7.2</b> 2. Exploratory data analysis (PCA &amp; heirarchical clustering) - identifying outliers and sources of variation in the data:</a></li>
<li class="chapter" data-level="9.7.3" data-path="codebook.html"><a href="codebook.html#run-deseq2"><i class="fa fa-check"></i><b>9.7.3</b> 3. Run DESeq2:</a></li>
<li class="chapter" data-level="9.7.4" data-path="codebook.html"><a href="codebook.html#check-the-fit-of-the-dispersion-estimates"><i class="fa fa-check"></i><b>9.7.4</b> 4. Check the fit of the dispersion estimates:</a></li>
<li class="chapter" data-level="9.7.5" data-path="codebook.html"><a href="codebook.html#create-contrasts-to-perform-wald-testing-on-the-shrunken-log2-foldchanges-between-specific-conditions"><i class="fa fa-check"></i><b>9.7.5</b> 5. Create contrasts to perform Wald testing on the shrunken log2 foldchanges between specific conditions:</a></li>
<li class="chapter" data-level="9.7.6" data-path="codebook.html"><a href="codebook.html#output-significant-results"><i class="fa fa-check"></i><b>9.7.6</b> 6. Output significant results:</a></li>
<li class="chapter" data-level="9.7.7" data-path="codebook.html"><a href="codebook.html#visualize-results-volcano-plots-heatmaps-normalized-counts-plots-of-top-genes-etc."><i class="fa fa-check"></i><b>9.7.7</b> 7. Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc.</a></li>
<li class="chapter" data-level="9.7.8" data-path="codebook.html"><a href="codebook.html#make-sure-to-output-the-versions-of-all-tools-used-in-the-de-analysis"><i class="fa fa-check"></i><b>9.7.8</b> 8. Make sure to output the versions of all tools used in the DE analysis:</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="codebook.html"><a href="codebook.html#functional-analysis-of-rna-seq-data"><i class="fa fa-check"></i><b>9.8</b> Functional analysis of RNA-seq data</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="codebook.html"><a href="codebook.html#over-representation-analysis-1"><i class="fa fa-check"></i><b>9.8.1</b> Over-representation analysis</a></li>
<li class="chapter" data-level="9.8.2" data-path="codebook.html"><a href="codebook.html#clusterprofiler-1"><i class="fa fa-check"></i><b>9.8.2</b> clusterProfiler</a></li>
<li class="chapter" data-level="9.8.3" data-path="codebook.html"><a href="codebook.html#gene-set-enrichment-analysis-gsea-1"><i class="fa fa-check"></i><b>9.8.3</b> Gene set enrichment analysis (GSEA)</a></li>
<li class="chapter" data-level="9.8.4" data-path="codebook.html"><a href="codebook.html#other-tools-and-resources-1"><i class="fa fa-check"></i><b>9.8.4</b> Other tools and resources</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNA-seq-analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dge-analysis-workflow" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> DGE analysis workflow<a href="dge-analysis-workflow.html#dge-analysis-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Differential expression analysis with DESeq2 involves multiple steps as displayed in the flowchart below in blue. Briefly, DESeq2 will:</p>
<div class="float">
<img src="img/DESeq2_workflow_2018.png" width="600" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<ul>
<li><p>normalize the raw counts using size factors to account for differences in library depth</p></li>
<li><p>estimate the gene-wise dispersions</p></li>
<li><p>shrink these estimates to generate more accurate dispersion estimates</p></li>
<li><p>fit the negative binomial (NG) model and perform hypothesis testing using the Wald test.</p></li>
</ul>
<p>This final step in the differential expression analysis workflow of fitting the raw counts to the NB model and performing the statistical test for differentially expressed genes, is the step we care about. This is the step that determines whether the mean expression levels of different sample groups are significantly different.</p>
<div class="float">
<img src="img/de_theory.png" width="600" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>The <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2 paper</a> was published in 2014, but the package is continually updated and available for use in R through Bioconductor.</p>
<div id="running-deseq2" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Running DESeq2<a href="dge-analysis-workflow.html#running-deseq2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Prior to performing the differential expression analysis, it is a good idea to know what <strong>sources of variation</strong> are present in your data, either by exploration during the QC and/or prior knowledge. Once you know the major sources of variation, you can remove them prior to analysis or control for them in the statistical model by including them in your <strong>design formula</strong>.</p>
<div id="design-formula" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Design formula<a href="dge-analysis-workflow.html#design-formula" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A design formula tells the statistical software the known sources of variation to control for, as well as, the factor of interest to test for during differential expression testing. For example, if you know that sex is a significant source of variation in your data, then sex should be included in your model. The design formula should have all of the factors in your metadata that account for major sources of variation in your data. The last factor entered in the formula should be the condition of interest.</p>
<p>For example, suppose you have the following metadata:</p>
<div class="float">
<img src="img/meta_example.png" width="300" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>If you want to examine the expression differences between treatments, and you know that major sources of variation include sex and age, then your design formula would be:</p>
<p><code>design &lt;- ~ sex + age + treatment</code></p>
<p>The tilde (<code>~</code>) should always proceed your factors and tells DESeq2 to model the counts using the following formula. Note the <strong>factors included in the design formula need to match the column names in the metadata</strong>.</p>
<hr />
<p><strong>Exercises</strong></p>
<ol style="list-style-type: decimal">
<li><p>Suppose you wanted to study the expression differences between the two age groups in the metadata shown above, and major sources of variation were <code>sex</code> and <code>treatment</code>, how would the design formula be written?</p></li>
<li><p>Based on our <strong>Mov10</strong> <code>metadata</code> dataframe, which factors could we include in our design formula?</p></li>
<li><p>What would you do if you wanted to include a factor in your design formula that is not in your metadata?</p></li>
</ol>
<hr />
</div>
<div id="mov10-de-analysis" class="section level3 hasAnchor" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> MOV10 DE analysis<a href="dge-analysis-workflow.html#mov10-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we know how to specify the model to DESeq2, we can run the differential expression pipeline on the <strong>raw counts</strong>.</p>
<p><strong>To get our differential expression results from our raw count data, we only need to run 2 lines of code!</strong></p>
<p>First we create a DESeqDataSet as we did in the <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html#2-create-deseq2-object">‘Count normalization’</a> lesson and specify the location of our raw counts and metadata, and input our design formula:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="dge-analysis-workflow.html#cb26-1" tabindex="-1"></a><span class="do">## Create DESeq object</span></span>
<span id="cb26-2"><a href="dge-analysis-workflow.html#cb26-2" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> data, <span class="at">colData =</span> meta, <span class="at">design =</span> <span class="sc">~</span> sampletype)</span></code></pre></div>
<p>Then, to run the actual differential expression analysis, we use a single call to the function <code>DESeq()</code>.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="dge-analysis-workflow.html#cb27-1" tabindex="-1"></a><span class="do">## Run analysis</span></span>
<span id="cb27-2"><a href="dge-analysis-workflow.html#cb27-2" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code></pre></div>
<p>By re-assigning the results of the function back to the same variable name (<code>dds</code>), we can fill in the <code>slots</code> of our <code>DESeqDataSet</code> object.</p>
<div class="float">
<img src="img/deseq_obj2.png" alt="deseq1" />
<div class="figcaption">deseq1</div>
</div>
<p><strong>Everything from normalization to linear modeling was carried out by the use of a single function!</strong> This function will print out a message for the various steps it performs:</p>
<pre><code>estimating size factors
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing</code></pre>
</div>
</div>
<div id="deseq2-differential-gene-expression-analysis-workflow" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> DESeq2 differential gene expression analysis workflow<a href="dge-analysis-workflow.html#deseq2-differential-gene-expression-analysis-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>With the 2 lines of code above, we just completed the workflow for the differential gene expression analysis with DESeq2. The steps in the analysis are output below:</p>
<div class="float">
<img src="img/deseq2_workflow_separate.png" width="200" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>We will be taking a detailed look at each of these steps to better understand how DESeq2 is performing the statistical analysis and what metrics we should examine to explore the quality of our analysis.</p>
<div id="step-1-estimate-size-factors" class="section level3 hasAnchor" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Step 1: Estimate size factors<a href="dge-analysis-workflow.html#step-1-estimate-size-factors" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The first step in the differential expression analysis is to estimate the size factors, which is exactly what we already did to normalize the raw counts.</p>
<div class="float">
<img src="img/deseq2_workflow_separate_sf.png" width="200" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>DESeq2 will automatically estimate the size factors when performing the differential expression analysis. However, if you have already generated the size factors using <code>estimateSizeFactors()</code>, as we did earlier, then DESeq2 will use these values.</p>
<p>To normalize the count data, DESeq2 calculates size factors for each sample using the <em>median of ratios method</em> discussed previously in the Count normalization (Chapter <a href="count-normalization.html#count-normalization">2</a>) lesson.</p>
<div id="mov10-de-analysis-examining-the-size-factors" class="section level4 hasAnchor" number="4.2.1.1">
<h4><span class="header-section-number">4.2.1.1</span> MOV10 DE analysis: examining the size factors<a href="dge-analysis-workflow.html#mov10-de-analysis-examining-the-size-factors" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s take a quick look at size factor values we have for each sample:</p>
<pre><code>## Check the size factors
sizeFactors(dds)

Mov10_kd_2 Mov10_kd_3 Mov10_oe_1 Mov10_oe_2 Mov10_oe_3 Irrel_kd_1 Irrel_kd_2 Irrel_kd_3 
 1.5646728  0.9351760  1.2016082  1.1205912  0.6534987  1.1224020  0.9625632  0.7477715  </code></pre>
<p>Take a look at the total number of reads for each sample:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="dge-analysis-workflow.html#cb30-1" tabindex="-1"></a><span class="do">## Total number of raw counts per sample</span></span>
<span id="cb30-2"><a href="dge-analysis-workflow.html#cb30-2" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">counts</span>(dds))</span></code></pre></div>
<p><em>How do the numbers correlate with the size factor?</em></p>
</div>
</div>
<div id="step-2-estimate-gene-wise-dispersion" class="section level3 hasAnchor" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Step 2: Estimate gene-wise dispersion<a href="dge-analysis-workflow.html#step-2-estimate-gene-wise-dispersion" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The next step in the differential expression analysis is the estimation of gene-wise dispersions. Before we get into the details, we should have a good idea about what dispersion is referring to in DESeq2.</p>
<div class="float">
<img src="img/deseq2_workflow_separate_dis.png" width="200" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p><strong>What is dispersion?</strong></p>
<p>Dispersion is a measure of spread or variability in the data. Variance, standard deviation, IQR, among other measures, can all be used to measure dispersion. However, DESeq2 uses a specific measure of dispersion (α) related to the mean (μ) and variance of the data: <code>Var = μ + α*μ^2</code>. For genes with moderate to high count values, the square root of dispersion will be equal to the coefficient of variation (<code>σ / μ</code>). So 0.01 dispersion means 10% variation around the mean expected across biological replicates.</p>
<p><strong>What does the DESeq2 dispersion represent?</strong></p>
<p>The DESeq2 dispersion estimates are <strong>inversely related to the mean</strong> and <strong>directly related to variance</strong>. <strong>Based on this relationship, the dispersion is higher for small mean counts and lower for large mean counts.</strong> The dispersion estimates for genes with the same mean will differ only based on their variance. <strong>Therefore, the dispersion estimates reflect the variance in gene expression for a given mean value.</strong></p>
<p>The plot of mean versus variance in count data below shows the variance in gene expression increases with the mean expression (each black dot is a gene). Notice that the relationship between mean and variance is linear on the log scale, and for higher means, we could predict the variance relatively accurately given the mean. However, <strong>for low mean counts, the variance estimates have a much larger spread; therefore, the dispersion estimates will differ much more between genes with small means</strong>.</p>
<div class="float">
<img src="img/deseq_mean_vs_variance.png" width="600" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p><strong>How does the dispersion relate to our model?</strong></p>
<p>To accurately model sequencing counts, we need to generate accurate estimates of within-group variation (variation between replicates of the same sample group) for each gene. With only a few (3-6) replicates per group, the <strong>estimates of variation for each gene are often unreliable</strong> (due to the large differences in dispersion for genes with similar means).</p>
<p>To address this problem, DESeq2 <strong>shares information across genes</strong> to generate more accurate estimates of variation based on the mean expression level of the gene using a method called ‘shrinkage’. <strong>DESeq2 assumes that genes with similar expression levels have similar dispersion.</strong></p>
<p><strong>Estimating the dispersion for each gene separately:</strong></p>
<p>The first step in estimating dispersion is to get the dispersion estimates for each gene. The dispersion for each gene is estimated using maximum likelihood estimation (MLE). MLE can be thought about as <strong>given the count values of the replicates, the most likely estimate of dispersion is calculated</strong>.</p>
</div>
<div id="step-3-fit-curve-to-gene-wise-dispersion-estimates" class="section level3 hasAnchor" number="4.2.3">
<h3><span class="header-section-number">4.2.3</span> Step 3: Fit curve to gene-wise dispersion estimates<a href="dge-analysis-workflow.html#step-3-fit-curve-to-gene-wise-dispersion-estimates" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The next step in the workflow is to fit a curve to the dispersion estimates for each gene. This curve represents the expected value of the dispersion given the expression values of the genes. The idea behind fitting a curve to the data is that different genes will have different scales of biological variability, but, over all genes, there will be a distribution of reasonable estimates of dispersion.</p>
<div class="float">
<img src="img/deseq2_workflow_separate_fit.png" width="200" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>This curve is displayed as a red line in the figure below, which plots the estimate for the <strong>expected dispersion value for genes of a given expression strength</strong>. Each black dot is a gene with an associated mean expression level and maximum likelihood estimation (MLE) of the dispersion (Step 1).</p>
<div class="float">
<img src="img/deseq_dispersion1.png" width="400" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
</div>
<div id="step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve" class="section level3 hasAnchor" number="4.2.4">
<h3><span class="header-section-number">4.2.4</span> Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve<a href="dge-analysis-workflow.html#step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The next step in the workflow is to shrink the gene-wise dispersion estimates toward the expected dispersion values to get the final dispersion estimates.</p>
<div class="float">
<img src="img/deseq2_workflow_separate_shr.png" width="200" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p>The curve allows for more accurate identification of differentially expressed genes when sample sizes are small, and the strength of the shrinkage for each gene depends on :</p>
<ul>
<li>how close gene dispersions are from the curve</li>
<li>sample size (more samples = less shrinkage)</li>
</ul>
<p><strong>This shrinkage method is particularly important to reduce false positives in the differential expression analysis.</strong> Genes with low dispersion estimates are shrunken towards the curve, and the more accurate, higher shrunken values are output for fitting of the model and differential expression testing.</p>
<p>Dispersion estimates that are slightly above the curve are also shrunk toward the curve for better dispersion estimation; however, genes with extremely high dispersion values are not. This is due to the likelihood that the gene does not follow the modeling assumptions and has higher variability than others for biological or technical reasons and therefore is likely to be an outlier. Shrinking the values toward the curve could result in false positives, so these values are not shrunken. These dispersion outliers are shown surrounded by blue circles below.</p>
<div class="float">
<img src="img/deseq_dispersion2.png" width="600" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<p><strong>This is a good plot to examine to ensure your data is a good fit for the DESeq2 model.</strong> You expect your data to generally scatter around the curve, with the dispersion decreasing with increasing mean expression levels. If you see a cloud or different shapes, then you might want to explore your data more to see if you have contamination (mitochondrial, etc.) or outlier samples. Note how much shrinkage you get across the whole range of means in the <code>plotDispEsts()</code> plot for any experiment with low degrees of freedom.</p>
<p>Examples of <strong>worrisome dispersion plots</strong> are shown below:</p>
<div class="float">
<img src="img/bad_dispersion1.png" width="600" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<div class="float">
<img src="img/bad_dispersion2.png" width="600" alt="Alt text" />
<div class="figcaption">Alt text</div>
</div>
<div id="mov10-de-analysis-exploring-the-dispersion-estimates-and-assessing-model-fit" class="section level4 hasAnchor" number="4.2.4.1">
<h4><span class="header-section-number">4.2.4.1</span> MOV10 DE analysis: exploring the dispersion estimates and assessing model fit<a href="dge-analysis-workflow.html#mov10-de-analysis-exploring-the-dispersion-estimates-and-assessing-model-fit" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s take a look at the dispersion estimates for our MOV10 data:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="dge-analysis-workflow.html#cb31-1" tabindex="-1"></a><span class="do">## Plot dispersion estimates</span></span>
<span id="cb31-2"><a href="dge-analysis-workflow.html#cb31-2" tabindex="-1"></a><span class="fu">plotDispEsts</span>(dds)</span></code></pre></div>
<p><img src="img/plotDispersion.png"></p>
<p><strong>Since we have a small sample size, for many genes we see quite a bit of shrinkage. Do you think our data are a good fit for the model?</strong></p>
<hr />

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="quality-control.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="model-and-hypothesis-testing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["RNA-seq-analysis.pdf", "RNA-seq-analysis.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
